### PR TITLE
adapter: accept parameters in /api/sql endpoint

### LINKED ISSUE #14067: sql: support parameters with SQL HTTP API
### Feature request

We aim to support parameterized queries with the `api/sql` HTTP endpoint.

We can support this by supporting a new `parameters` key in the request's body, which takes an array of values of the form:

```json
{
  "type": "<resolvable type name>",
  "value": "<JSON-ified value>"
}
```

In the `simple_execute` method, we can then generate the appropriate values, and bind them to a portal. We currently use the empty portal for these requests, but can instead use an "anonymous" portal named after a UUID, the name of which we will need to store only for the length of the request (the entirety of which we receive).

### PR DESCRIPTION
**Pull request by sploiselle**
Accept user parameters in a PG Simple Query-friendly way.

This approach is inspired by PG, namely this sentence
from the description of the Extended Query protocol:

> Furthermore, additional features are available [in Extended Query],
> such as the possibility of supplying data values as separate
> parameters instead of having to insert them directly into a query
> string.

This design does a slightly more safe version of what PG suggests,
by using a visitor over the AST and inserting the cast expressions
wherever there are parameters. Surely this is what PG themselves
must do.

For details on this design, see:
https://github.com/MaterializeInc/materialize/pull/14073

### Motivation

This PR adds a known-desirable feature.  Closes #14067

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Support parameters over the `api/sql` endpoint using a structure like:
     ```json
    {
      "sql": "INSERT INTO foo VALUES ($1);",
      "parameters": [{"type": "int", "value": "1"}]
    }
    ```

### CODE REVIEW

**Comment by sploiselle** [comment]
@mjibson just wanted to ping you on this for review--ty

File: src/environmentd/tests/server.rs [hunk 1]
```diff
@@ -132,64 +132,238 @@ fn test_http_sql() -> Result<(), Box<dyn Error>> {
...
+        TestCase {
+            query: "select length($1), length($2)",
+            params: Some(vec![("sum(a)", "text"), ("SELECT * FROM t;", "text")]),
+            status: StatusCode::OK,
+            body: r#"{"results":[{"rows":[[6,16]],"col_names":["length","length"]}]}"#,
+        },
+        // Too many parameters (OK)
+        TestCase {
+            query: "select $1 as col",
+            params: Some(vec![("1", "int"), ("2", "int8")]),
```
Comment (mjibson) [inline src/environmentd/tests/server.rs:265]: Why is this ok?

File: src/adapter/src/client.rs [hunk 2]
```diff
@@ -397,13 +400,79 @@ impl SessionClient {
...
+                            None => {
+                                self.error = Some(anyhow!("there is no parameter ${}", i));
+                                return;
+                            }
+                        };
+
+                        match mz_sql_parser::parser::parse_expr(&format!(
+                            "CAST ('{}' AS {})",
+                            display::escape_single_quote_string(val),
+                            typ
```
Comment (mjibson) [inline src/adapter/src/client.rs:437]: Using the `typ` as passed by the user without validating it is a means for SQL injection if we use string formatting.

File: src/adapter/src/client.rs [hunk 3]
```diff
@@ -397,13 +400,79 @@ impl SessionClient {
...
+        }
+
         let stmts =
             mz_sql::parse::parse(&stmts).map_err(|e| AdapterError::Unstructured(e.into()))?;
         let num_stmts = stmts.len();
         const EMPTY_PORTAL: &str = "";
         let mut results = vec![];
-        for stmt in stmts {
+        for mut stmt in stmts {
+            inline_params(&params, &mut stmt).map_err(AdapterError::Unstructured)?;
```
Comment (mjibson) [inline src/adapter/src/client.rs:474]: We should use the actual param stuff instead of inlining with strings. Can you use the same logic that the extended protocol does? The `parse`, `bind`, `execute` fns in protocol.rs have all the logic needed to do this correctly.

**Comment by mjibson** [comment]
Eep sorry. Had looked at this but didn't complete and forgot about it. Thanks for the reminder.

File: src/adapter/src/client.rs (hunk 2 above)
Comment (sploiselle) [reply src/adapter/src/client.rs:437]: It is validated; the second argument can only be parsed as a data type. Do you have any suggestions for a means by which this could be injected into?

File: src/adapter/src/client.rs (hunk 3 above)
Comment (sploiselle) [reply src/adapter/src/client.rs:474]: This is how PG handles parameters in the simple protocol. Using portals, etc. turns this into some hybrid version of the simple and extended protocols, and adds more layers of complexity. Is there a concern with this implementation?

File: src/environmentd/tests/server.rs (hunk 1 above)
Comment (sploiselle) [reply src/environmentd/tests/server.rs:265]: Ah--do you have any reference where unused parameters should generate errors? It's trivial to implement, but I couldn't find anything suggesting they should, but might very well be mistaken.

File: src/adapter/src/client.rs (hunk 3 above)
Comment (mjibson) [reply src/adapter/src/client.rs:474]: PG doesn't handle parameters in the simple protocol. Where did you see that?

File: src/adapter/src/client.rs (hunk 2 above)
Comment (mjibson) [reply src/adapter/src/client.rs:437]: I had in mind something like `int) ...`. Can we explicitly parse this as a type and then inject the `to_string` of that here?

File: src/environmentd/tests/server.rs (hunk 1 above)
Comment (mjibson) [reply src/environmentd/tests/server.rs:265]:
```
send
Parse {"query": "SELECT $1 + 1"}
Bind {"values": ["1", "2"]}
Sync
----

until
ReadyForQuery
----
ParseComplete
ErrorResponse {"fields":[{"typ":"S","value":"ERROR"},{"typ":"C","value":"08P01"},{"typ":"M","value":"bind message supplies 2 parameters, but prepared statement \"\" requires 1"}]}
ReadyForQuery {"status":"I"}
```

Here's a pgtest that generates an error against postgres. We don't currently test this but we should.

File: src/adapter/src/client.rs (hunk 3 above)
Comment (sploiselle) [reply src/adapter/src/client.rs:474]: Ah, I had misunderstood the PG documentation w/r/t parameter in the Simple Query API. Bad point.

However, I still don't think inlining the extended protocol approach is right here, unless we want to move all of the requests that use parameters to instead be processed that way?

Some of the things that threw me off of that approach:
- `StateMachine::parse` takes strings and only permits single statements to be executed; this means we'd have to take the input to Vec<statements>, and then take each of those individual statements back to strings to be parsed. That smells fishy to me.
-  `StateMachine::parse` requires inputting the `param_oid`s, but `simple_execute` doesn't have catalog access, so we've have to perform some complex queries to try and infer the correct OID for the type name--name resolution is complex and trying to manifest that in the HTTP API seems very challenging.

Doing an AST rewrite accomplishes the same thing simply and safely, so I'm very apprehensive about trying to bring the `parse`/`bind`/`execute`  code into this. If you'd like to spend some time with me showing me how you'd approach it, though, very glad to set up some time.

Comment (sploiselle) [reply src/adapter/src/client.rs:474]: For posterity, I under misunderstood Matt's suggestion; he was not talking about using those functions but instead using the logic they contain.

@mjibson can you summarize why you think this approach is superior to the AST rewrite? It seems like it's introducing a good bit of complexity (e.g. bringing in the junk from `bind`) and am trying to figure out what it gets us.

Comment (mjibson) [reply src/adapter/src/client.rs:474]: This gets us parity with the existing extended protocol which will allow for some possible code reuse, and avoids adding a new method that uses string formatting which has dubious security properties (hard to prove it doesn't have a SQL injection that we're not aware of).

Comment (sploiselle) [reply src/adapter/src/client.rs:474]: I'll give you that modeling things after the extended protocol is a matter of taste. However, we can trivially ensure that the received value, when parsed as an expression, is an `ast::defs::value::Value::String`.

```rust
let val = match mz_sql_parser::parser::parse_expr(
    display::escape_single_quote_string(val),
) {
    Ok(r @ mz_sql_parser::ast::defs::Value::String(..)) => r,
    _ => bail!("not a string!"),
};
```

**Comment by sploiselle** [comment]
Closing in favor of #14683
